packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -125,28 +302,105 @@ export const cacheUpdateMatrix: CacheUpdateMatrix<Client> = {
-            // TODO: Rather than invalidate, can we selectively update?
-            invalidate: [{queryKey: getCustomerProductLists.queryKey(parameters)}]
+            update: [
+                {
+                    queryKey: getCustomerProductList.queryKey(parameters)
+                },
+                {
+                    queryKey: getCustomerProductLists.queryKey(parameters),
+                    updater: createUpdateFunction((result: CustomerProductListResult) => {
+                        const listIndex = result.data.findIndex(({id}) => id === response.id)
+
+                        // Return undefined if we didn't find the product list we were looking for.
+                        if (listIndex < 0) {
+                            return
+                        }
+
+                        // Update the product list.
+                        result.data[listIndex] = response
+
+                        return result
+                    })
+                }
+            ]
         }
     },
-    updateCustomerProductListItem(customerId, {parameters}) {
+    updateCustomerProductListItem(customerId, {parameters}, response) {
         return {
-            update: [{queryKey: getCustomerProductListItem.queryKey(parameters)}],
-            // TODO: Rather than invalidate, can we selectively update?
-            invalidate: [
-                {queryKey: getCustomerProductList.queryKey(parameters)},
-                {queryKey: getCustomerProductLists.queryKey(parameters)}
+            update: [
+                {
+                    queryKey: getCustomerProductListItem.queryKey(parameters)
+                },
+                {
+                    queryKey: getCustomerProductList.queryKey(parameters),
+                    updater: createUpdateFunction((list: CustomerProductList) => {
+                        // Find the index of the item we want to update.
+                        const itemIndex = list.customerProductListItems?.findIndex(
+                            ({id}) => id === parameters.itemId
+                        )
+
+                        // Return undefined when item isn't found.
+                        if (typeof itemIndex === 'undefined' || itemIndex < 0) {
+                            return
+                        }
+
+                        list.customerProductListItems![itemIndex] = response
+
+                        return list
+                    })
+                },
+                {
+                    queryKey: getCustomerProductLists.queryKey(parameters),
+                    updater: createUpdateFunction((result: CustomerProductListResult) => {
+                        // Find the list with the current list id.
+                        const listIndex = result.data.findIndex(({id}) => id === parameters.listId)
+                        // Find the index of the item in the list.
+                        const itemIndex = result?.data[
```

Comment:
```suggestion
                        const itemIndex = result?.data[
```
Noticed you still have a lot of `result?.` and `customer?.` floating around, even though those checks are already handled by `createUpdateFunction`.